File: Analise/heap.js
class Heap {
  constructor(vetor = []) {
    this.heap = [];
    for (let i = 0; i < vetor.length; i++) {
      this.insert(vetor[i]);
    }
  }

  insert(elemento) {
    this.heap.push(elemento);
    this.heapify_up(this.heap.length - 1);
  }

  make_heap(vetor) {
    for (let i = 0; i < vetor.length; i++) {
      this.insert(vetor[i]);
    }
  }

  get_max() {
    if (this.heap.length === 0) {
      return null;
    }
    return this.heap[0];
  }

  remove_max() {
    const max = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapify_down(0);
    return max;
  }

  heapify_up(index) {
    let index_atual = index;
    while (index_atual > 0) {
      const index_pai = Math.floor((index_atual - 1) / 2);
      if (this.heap[index_atual] > this.heap[index_pai]) {
        [this.heap[index_atual], this.heap[index_pai]] = [this.heap[index_pai], this.heap[index_atual]];
        index_atual = index_pai;
      } else {
        break;
      }
    }
  }

  heapify_down(index) {
    let index_atual = index;
    const n = this.heap.length;
    while (true) {
      let index_esq_filho = 2 * index_atual + 1;
      let index_dir_filho = 2 * index_atual + 2;
      let index_pai = index_atual;

      if (index_esq_filho < n && this.heap[index_esq_filho] > this.heap[index_pai]) {
        index_pai = index_esq_filho;
      }
      if (index_dir_filho < n && this.heap[index_dir_filho] > this.heap[index_pai]) {
        index_pai = index_dir_filho;
      }

      if (index_pai !== index_atual) {
        [this.heap[index_atual], this.heap[index_pai]] = [this.heap[index_pai], this.heap[index_atual]];
        index_atual = index_pai;
      } else {
        break;
      }
    }
  }

  heap_to_vetor() {
    return [...this.heap];
  }
}

function encontrar_maiores_elementos(v, tamanho_amostra) {
  const heap = new Heap();
  for (const subvetor of v) {
    for (const elemento of subvetor) {
      heap.insert(elemento);
    }
  }
  const vetor_solucao = [];
  for (let i = 0; i < tamanho_amostra && heap.heap.length > 0; i++) {
    vetor_solucao.push(heap.remove_max());
  }
  return vetor_solucao;
}







/*

const teste = [ [ 1, 4, 2 ], [ 5, 5, 1 ], [ 3, 2, 5 ], [ 2 ] ]
const tamanho_amostra = 10;



const vetor_solucao = encontrar_maiores_elementos(teste, tamanho_amostra);
console.log("Vetor solução:", vetor_solucao);
*/


module.exports={Heap,encontrar_maiores_elementos}

File: Projeto II/Heap.py
class Heap:
    def __init__(self, vetor=[]):
        self.heap = []
        for elemento in vetor:
            self.insert(elemento)

    def insert(self, elemento):
        self.heap.append(elemento)
        self.heapify_up(len(self.heap) - 1)

    def make_heap(self, vetor):
        for elemento in vetor:
            self.insert(elemento)

    def get_max(self):
        if len(self.heap) == 0:
            return None
        return self.heap[0]

    def remove_max(self):
        if len(self.heap) == 0:
            return None

        max_element = self.heap[0]
        self.heap[0] = self.heap.pop()
        self.heapify_down(0)

        return max_element

    def heapify_up(self, index):
        index_atual = index
        while index_atual > 0:
            index_pai = (index_atual - 1) // 2
            if self.heap[index_atual] > self.heap[index_pai]:
                self.heap[index_atual], self.heap[index_pai] = self.heap[index_pai], self.heap[index_atual]
                index_atual = index_pai
            else:
                break

    def heapify_down(self, index):
        index_atual = index
        n = len(self.heap)
        while True:
            index_esq_filho = 2 * index_atual + 1
            index_dir_filho = 2 * index_atual + 2
            index_pai = index_atual

            if index_esq_filho < n and self.heap[index_esq_filho] > self.heap[index_pai]:
                index_pai = index_esq_filho

            if index_dir_filho < n and self.heap[index_dir_filho] > self.heap[index_pai]:
                index_pai = index_dir_filho

            if index_pai != index_atual:
                self.heap[index_atual], self.heap[index_pai] = self.heap[index_pai], self.heap[index_atual]
                index_atual = index_pai
            else:
                break

    def heap_to_vetor(self):
        return self.heap[:]


def encontrar_maiores_elementos(v, vetor_solucao):
    i = 0
    l = v
    while len(l) > 0:
        maior_elemento_sub_vetor = []

        vetor = l
        l = []
        print("\n\n")
        print("vetor", vetor, "i=", i)
        for sub_vetor in vetor:
            heap_pri = Heap()
            heap_pri.make_heap(sub_vetor)

            maior_elemento_sub_vetor.append(heap_pri.get_max())

            heap_pri.remove_max()

            vetor = heap_pri.heap_to_vetor()
            if len(vetor) == 0:
                continue
            else:
                l.append(vetor)
            print("l dentro do for ", l)

        heap_seg = Heap()
        heap_seg.make_heap(maior_elemento_sub_vetor)
        maior_de_todos = heap_seg.get_max()
        print("Heap_seg", heap_seg)
        vetor_solucao.append(maior_de_todos)
        heap_seg.remove_max()
        aux_vetor = heap_seg.heap_to_vetor()
        print(">>aux_vetor", aux_vetor)

        for j in range(len(aux_vetor)):
            print("aux_vetor", aux_vetor[j], "j", j)

            print("lll", l)
            l[j].extend([aux_vetor[j]])
            print("lll", l)

        print("l", l)

        print("vetor_solucao", vetor_solucao)

        i += 1
        if i == 10:
            break


# teste [16,9,7],[6,6,6],[4,2,1]
""" 
teste = [[9, 4, 6], [7, 16, 6], [2, 1, 6]]
vetor_solucao = []

encontrar_maiores_elementos(teste, vetor_solucao)
print("Vetor solução:", vetor_solucao)
print("Teste final:", teste)
""" 